### PR TITLE
Add Remote Playback as an alternative

### PR DESCRIPTION
**Pull request by takumif**
Mark: PTAL, thanks!

### CODE REVIEW

File: explainer.md [hunk 1]
```diff
@@ -76,6 +76,8 @@ page they switch the slide deck to presentation mode, and select the smart
 projector as the target.  The presentation's audio and video are mirrored to the
 smart projector.
 
+## Detailed design discussion
+
 ### Sample Code
```
Comment (markafoltz) [inline explainer.md:85]: Do you mind mentioning that the first design alternative extends the Presentation API before diving into the sample code for it?

File: explainer.md [hunk 2]
```diff
@@ -169,6 +169,75 @@ time.
...
+currently only supports the playback of media elements, it would need to be 
+extended to allow the remote playback of entire tab contents.
+
+Arguments in favor of extending the Remote Playback API:
+1. Sending a mirroring stream is arguably conceptually closer to Remote 
+Playback than Presentation, given the latter is about setting up a separate 
+receiver page and bidirectional communication between the controller and the 
+receiver pages.
+1. Configuring latency and audio playback does not apply to the existing use of 
+Presentation API, but it can for the existing use of Remote Playback.
```
Comment (markafoltz) [inline explainer.md]: This statement depends on the implementation of Remote Playback; for example, if it's flinging the media so it's only played remotely, then latency/audio may not be configurable.

File: explainer.md [hunk 3]
```diff
@@ -169,6 +169,75 @@ time.
...
+
+Arguments against it:
+1. In the existing Remote Playback, we halt the local playback when it gets
```
Comment (markafoltz) [inline explainer.md]: During Remote Playback, we ...

File: explainer.md [hunk 4]
```diff
@@ -169,6 +169,75 @@ time.
...
+1. In the existing Remote Playback, we halt the local playback when it gets 
+activated, so that'd be inconsistent with self-mirroring, in which playback 
+happens simultaneously on both displays.
+1. Remote Playback is intended for media playback, and mirroring a page 
+is arguably conceptually different.
+
+Some implications of extending the Remote Playback API:
+1. It would mean that the user agent would let the page know whenever it’s 
+being tab mirrored (via `document.remote.onconnect`), even when mirroring is 
+initiated via the user agent's UI.
```
Comment (markafoltz) [inline explainer.md]: Would this necessarily be true?   If it is, it could lead to negative effects on the user experience.

File: explainer.md [hunk 5]
```diff
@@ -169,6 +169,75 @@ time.
...
+document.querySelector('#mirrorBtn').addEventListener('click', () => {
+  document.remote.prompt();
+});
+
+document.remote.addEventListener('connect', event => {
+  if (event.target.captureParams) {
+    document.remote.captureParams.latencyHint = 'low';
+    document.remote.captureParams.audioPlayback = 'local';
+  }
+});
```
Comment (markafoltz) [inline explainer.md:226]: Did you have an idea on how to adjust the configuration after mirroring has started?

Comment (markafoltz) [reply explainer.md:226]: Or would that be done on the document.remote object?

File: explainer.md [hunk 6]
```diff
@@ -169,6 +169,75 @@ time.
...
+
+Some implications of extending the Remote Playback API:
+1. It would mean that the user agent would let the page know whenever it’s 
+being tab mirrored (via `document.remote.onconnect`), even when mirroring is 
+initiated via the user agent's UI.
+1. If both `document.remote.remote` and `navigator.presentation.defaultRequest` 
+are set, and the user agent's secondary display picker is shared between the 
+Remote Playback and Presentation APIs, one API would need to be chosen as the 
+default when the picker is shown through the user agent's UI (i.e. not through 
+`RemotePlayback#prompt()` or `PresentationRequest#start()`).
```
Comment (markafoltz) [inline explainer.md]: There was a proposal at TPAC 2019 to allow multiple APIs to share the result of secondary screen selection:

https://www.w3.org/2019/09/16-webscreens-minutes.html#x04

Could that help address this issue?

File: explainer.md [hunk 7]
```diff
@@ -169,6 +169,75 @@ time.
...
+```
+partial interface Document {
+  attribute RemotePlayback remote;
+};
+
+partial interface RemotePlayback {
+  attribute CaptureParameters? captureParams;
+};
+
+// The below is the same as in the Presentation API proposal.
```
Comment (markafoltz) [inline explainer.md]: It should be okay to omit the full definitions if they are identical; that makes them easier to update in future iterations.

**Review by markafoltz — APPROVED** [review]
Thanks for this contributions!  A few questions/suggestions inline, but feel free to merge whenever you're ready.

File: explainer.md (hunk 5 above)
Comment (takumif) [reply explainer.md:226]: Right, the `document.remote.captureParams` fields can be changed mid-session.

File: explainer.md (hunk 4 above)
Comment (takumif) [reply explainer.md]: I think we'd want to notify the page when tab mirroring is started via the browser's UI, to allow the page to keep the behavior consistent with mirroring started by the page (just like a Presentation started via browser UI notifies the page). 

What negative effects on UX do you anticipate?

File: explainer.md (hunk 6 above)
Comment (takumif) [reply explainer.md]: IIRC we didn't come to a conclusion on how the UA would choose one API over the other, so I think it was also an open issue there. Added a link to the discussion.

**Review by takumif — COMMENTED** [review]
Thanks for the comments, I've addressed them. AFAICT I don't have the rights to merge the PR, could you merge it for me?

**Review by markafoltz — APPROVED** [review]

**Comment by markafoltz** [comment]
Sorry for the delay - I recently updated my email filters so I don't miss notifications for PRs like these.

**Pull request merged**